protocols/relay/examples/relay.rs

```diff
@@ -82,3 +154,149 @@ fn main() -> Result<(), Box<dyn Error>> {
         Poll::Pending
     }))
 }
+
+fn publish_listener_peer(addr: &libp2p::Multiaddr, mode: &Mode, local_peer_id: PeerId) -> () {
+    match mode {
+        Mode::Relay => {
+            let address = format!("{}/p2p/{}/p2p-circuit", addr, local_peer_id);
+            set_shared_data(RELAY_ADDRESS_KEY, &address);
+        }
+        Mode::ClientListen => {
+            let address = format!("/p2p/{}/{}", addr, local_peer_id);
+            set_shared_data(CLIENT_LISTEN_ADDRESS_KEY, &address);
+        }
+        Mode::ClientDial => {}
+    }
+}
+
+fn generate_ed25519(deterministic_peer_secret: u8) -> identity::Keypair {
+    let mut bytes = [0u8; 32];
+    bytes[0] = deterministic_peer_secret;
+
+    let secret_key = ed25519::SecretKey::from_bytes(&mut bytes)
+        .expect("this returns `Err` only if the length is wrong; the length is correct; qed");
+    identity::Keypair::Ed25519(secret_key.into())
+}
+
+/// Get the address for relay mode
+fn get_relay_address(opt: &Opt) -> String {
+    match &opt.address {
+        Some(address) => address.clone(),
+        None => {
+            println!("--address argument was not provided, will use the default listening relay address: {}",DEFAULT_RELAY_ADDRESS);
+            DEFAULT_RELAY_ADDRESS.to_string()
+        }
+    }
+}
+
+/// Get the address for client_listen mode
+fn get_relay_peer_address(opt: &Opt) -> String {
+    match &opt.address {
+        Some(address) => address.clone(),
+        None => {
+            match get_shared_data(RELAY_ADDRESS_KEY) {
+                Some(address) => address,
+                None => panic!("Please provide relayed listen address such as: <addr-relay-server>/p2p/<peer-id-relay-server>/p2p-circuit"),
+            }
+        }
+    }
+}
+
+/// Get the address for client-dial mode
+fn get_client_listen_address(opt: &Opt) -> String {
+    match &opt.address {
+        Some(address) => address.clone(),
+        None => {
+            match get_shared_data(CLIENT_LISTEN_ADDRESS_KEY) {
+                Some(address) => address.to_string(),
+                None => panic!("Please provide client listen address such as: <addr-relay-server>/p2p/<peer-id-relay-server>/p2p-circuit/p2p/<peer-id-listening-relay-client>"),
+            }
+        }
+    }
+}
+
+/// Placeholder to publish the peer address based on given key
+fn set_shared_data(key: &str, value: &str) -> () {
+    println!(
+        "TODO publish to KAD or discovery service ... Set the key: {} data to:{}",
+        key, value
+    );
+}
```

Comment:
```suggestion
```

I would prefer rust-libp2p not to contain any placeholder code. Could you remove everything around some external discovery mechanism?